### ISSUE TITLE
Broken HTTP parser and matcher

### ISSUE THREAD

**Issue by krizhanovsky** [tempesta-tech/tempesta#1474]
## Scope

`TlsHandshakeTest.test_10byte_transfer` is broken for a while. The test just sends a short HTTP request in 10-bytes chunks and then matches it against HTTPtables. The sequence of decrypted chunks: `GET / H`, `TTP/1.1\r\nH`, `ost: tempe`, `sta-tech.c`, `om\r\n\r\n`.

This patch for Tempesta FW
```
--- a/tempesta_fw/http_match.c
+++ b/tempesta_fw/http_match.c
@@ -121,6 +121,7 @@ hdr_val_eq(const TfwHttpReq *req, const TfwHttpMatchRule *rule,
        hdr = &req->h_tbl->tbl[id];
        if (TFW_STR_EMPTY(hdr))
                return false;
+       tfw_str_dprint(hdr, "AK_DBG full header");
 
        if (op == TFW_HTTP_MATCH_O_WILDCARD)
                return true;
@@ -134,6 +135,7 @@ hdr_val_eq(const TfwHttpReq *req, const TfwHttpMatchRule *rule,
         */
        flags |= TFW_STR_EQ_CASEI;
 
+       tfw_str_dprint(&hdr_val, "AK_DBG header value");
        if (op == TFW_HTTP_MATCH_O_SUFFIX)
                return tfw_str_eq_cstr_off(&hdr_val, hdr_val.len - str_len,
                                           str, str_len, flags);
diff --git a/tempesta_fw/str.c b/tempesta_fw/str.c
index 5ae83755..711afdea 100644
--- a/tempesta_fw/str.c
+++ b/tempesta_fw/str.c
@@ -1041,6 +1041,8 @@ tfw_str_eq_cstr(const TfwStr *str, const char *cstr, int cstr_len,
                BUG_ON(chunk->len &&  !chunk->data);
 
                len = min(clen, (int)chunk->len);
+               T_DBG("AK_DBG match flags=%x [%.*s] against [%.*s]\n",
+                       flags, len, cstr, len, chunk->data);
                if (cmp(cstr, chunk->data, len))
                        return false;
```
produces trace
```
[50906.351210] [tempesta fw]   AK_DBG full header: addr=ffff993332462348 skb=ffff993332ecbc00 len=19 flags=2 eolen=2:
[50906.353193] [tempesta fw]     duplicate ffff993332462348, len=19, flags=2 eolen=2:
[50906.354984] [tempesta fw]      len=1, eolen=0 ptr=ffff99335c57a309 flags=0 'H'
[50906.356636] [tempesta fw]      len=1, eolen=0 ptr=ffff993332932304 flags=100 ' '
[50906.358191] [tempesta fw]      len=5, eolen=0 ptr=ffff993332932305 flags=0 'tempe'
[50906.359762] [tempesta fw]      len=10, eolen=0 ptr=ffff99334f393300 flags=0 'sta-tech.c'
[50906.361500] [tempesta fw]      len=2, eolen=0 ptr=ffff99335c51a300 flags=0 'om'
[50906.364215] [tempesta fw]   AK_DBG Host header value: addr=ffff99337fc038a0 skb=ffff993332ecbc00 len=12 flags=2 eolen=2:
[50906.366617] [tempesta fw]     duplicate ffff99337fc038a0, len=12, flags=2 eolen=2:
[50906.368274] [tempesta fw]      len=10, eolen=0 ptr=ffff99334f393300 flags=0 'sta-tech.c'
[50906.369977] [tempesta fw]      len=2, eolen=0 ptr=ffff99335c51a300 flags=0 'om'
[50906.371667] [tempesta fw]   AK_DBG match flags=2 [tempesta-t] against [sta-tech.c]
```
So not only the parser loses `ost:` chunk, but also `tfw_http_msg_clnthdr_val()` selects wrong `TfwStr` subsequence.

## Testing

The bug is produced by existing `TlsHandshakeTest.test_10byte_transfer`. Maybe the parser unit test should be adjusted.

**Comment by ikoveshnikov** [comment]
Fixed in 87e44b41c7edb64637522d9a6f79453dd4b1d707

Seems I've faced the same issue in #1458. The commit  fixes `ost:` chunk loss and the test works just fine for me. I have also added more unit tests on the host header parsing, so we shouldn't get into this issue any more.

**Comment by krizhanovsky** [comment]
I confirm that the bug is fixed in the `ik-1054-update` branch

**Comment by ikoveshnikov** [comment]
Fixed in 87e44b4
